zipline/data/data_portal.py

```diff
@@ -463,41 +463,76 @@ def get_spot_value(self, asset, field, dt, data_frequency):
-                return None
-
-        if data_frequency == "daily":
-            if field == "contract":
-                return self._get_current_contract(asset, session_label)
-            else:
-                return self._get_daily_spot_value(asset, field, session_label)
-        else:
-            if field == "last_traded":
-                return self.get_last_traded_dt(asset, dt, 'minute')
-            elif field == "price":
-                return self._get_minute_spot_value(asset, "close", dt,
-                                                   ffill=True)
-            elif field == "contract":
-                return self._get_current_contract(asset, dt)
+        for asset in assets:
+            if self._is_extra_source(
+                    asset, field, self._augmented_sources_map):
+                spot_values.append(self._get_fetcher_value(asset, field, dt))
+                continue
+
+            if field not in BASE_FIELDS:
+                raise KeyError("Invalid column: " + str(field))
+
+            session_label = self.trading_calendar.minute_to_session_label(dt)
+
+            if dt < asset.start_date or \
+                    (data_frequency == "daily" and
+                        session_label > asset.end_date) or \
+                    (data_frequency == "minute" and
+                     session_label > asset.end_date):
+                if field == "volume":
+                    spot_values.append(0)
+                    continue
+                elif field != "last_traded":
+                    spot_values.append(np.NaN)
+                    continue
+                elif field == "contract":
+                    spot_values.append(None)
+                    continue
+
+            if data_frequency == "daily":
+                if field == "contract":
+                    spot_values.append(
+                        self._get_current_contract(asset, session_label)
+                    )
+                else:
+                    spot_values.append(
+                        self._get_daily_spot_value(asset, field, session_label)
+                    )
             else:
-                return self._get_minute_spot_value(asset, field, dt)
+                if field == "last_traded":
+                    spot_values.append(
+                        self.get_last_traded_dt(asset, dt, 'minute')
+                    )
+                elif field == "price":
+                    spot_values.append(
+                        self._get_minute_spot_value(
+                            asset, "close", dt, ffill=True,
+                        )
+                    )
+                elif field == "contract":
+                    spot_values.append(self._get_current_contract(asset, dt))
+                else:
+                    spot_values.append(
+                        self._get_minute_spot_value(asset, field, dt)
+                    )
+
+        return spot_values[0] if assets_is_scalar else spot_values
```

Comment:
I have not profiled it, but always creating the out array may be a place to watch for a performance regression if/when `get_spot_value` is called in a tight loop.

A good way to check is to profile a 'buy and hold' type algorithm over a 500+ sized universe to make sure that that the the position tracker checks for the end of day values does not cause too much of a regression.
(Also the position tracker could be changed to pass a vector of assets, to take advantage of this change.)